test/network/network_test.cpp

```diff
@@ -247,31 +247,87 @@ TEST_F(NetworkTests, LargePacketsTest) {
+    EXPECT_NE(io_socket_unique_ptr, nullptr);
+    io_sockets.emplace_back(std::move(io_socket_unique_ptr));
+  }
+
+  for (auto &socket : io_sockets) {
+    ManualPacketUtil::TerminateConnection(socket->GetSocketFd());
+    socket->Close();
+  }
+
+  for (size_t i = 0; i < CONNECTION_THREAD_COUNT * 2; i++) {
+    auto io_socket_unique_ptr = network::ManualPacketUtil::StartConnection(port_);
+    EXPECT_NE(io_socket_unique_ptr, nullptr);
+    io_sockets.emplace_back(std::move(io_socket_unique_ptr));
+  }
+  io_sockets.clear();
+
+  for (auto &socket : io_sockets) {
+    ManualPacketUtil::TerminateConnection(socket->GetSocketFd());
+    socket->Close();
+  }
+}
+
 /**
  * This is meant to overload the network layer with multiple concurrent client threads. It was made to uncover
  * a bug where ConnectionHandlerTask had a few race conditions amongst its fields. Two threads using the same
  * ConnectionHandlerTask instance could corrupt its fields.
  */
 // NOLINTNEXTLINE
 TEST_F(NetworkTests, RacerTest) {
-  std::vector<std::thread> threads;
-  threads.reserve(CONNECTION_THREAD_COUNT * 2);
-  for (size_t i = 0; i < CONNECTION_THREAD_COUNT * 2; i++) {
-    threads.emplace_back([=] {
-      try {
-        auto io_socket_unique_ptr = network::ManualPacketUtil::StartConnection(port_);
-        auto io_socket = common::ManagedPointer(io_socket_unique_ptr);
-        ManualPacketUtil::TerminateConnection(io_socket->GetSocketFd());
-        io_socket->Close();
-      } catch (const std::exception &e) {
-        EXPECT_TRUE(false);
-      }
-    });
-  }
-  for (auto &t : threads) {
-    t.join();
+  //  due to issue #766 PostgresProtocolHandler cannot consistently handle
+  //  simultaneous client requests so for now we are allowing for some failure tolerance
+
+  // The number of threads to use in each trial
+  std::vector<size_t> thread_counts = {
+      // clang-format off
+      2,
+      CONNECTION_THREAD_COUNT,
+      CONNECTION_THREAD_COUNT * 2,
+      CONNECTION_THREAD_COUNT * 3
+      // clang-format on
+  };
+
+  for (auto num_threads : thread_counts) {
+    std::vector<std::thread> threads;
+    threads.reserve(num_threads);
+    NETWORK_LOG_INFO("Number of threads: {0}", num_threads);
+
+    std::atomic_int successes = 0;
+    for (size_t i = 0; i < num_threads; i++) {
+      threads.emplace_back([this, &successes] {
+        try {
+          auto io_socket_unique_ptr = network::ManualPacketUtil::StartConnection(port_);
+          if (io_socket_unique_ptr == nullptr) {
+            return;
+          }
+          successes++;
+          auto io_socket = common::ManagedPointer(io_socket_unique_ptr);
+          ManualPacketUtil::TerminateConnection(io_socket->GetSocketFd());
+          io_socket->Close();
+        } catch (const std::exception &e) {
+          EXPECT_TRUE(false);
+        }
+      });
+    }
+    for (auto &t : threads) {
+      t.join();
+    }
+    threads.clear();
+    EXPECT_EQ(successes, num_threads);
```

Comment:
This test code is *amazing*. I have never seen a test case like this before. It is mind blowing.